glommio/src/io/dma_file_stream.rs

```diff
@@ -904,55 +906,73 @@ impl DmaStreamWriterState {
-                self.out_of_order_write_returns.pop();
-            } else {
-                return;
+    fn on_flush_success(&mut self, flush_pos: u64) {
+        let mut complete_prefix_len = 0usize;
+        let mut complete_prefix_finalized = false;
+        let mut found_index = None;
+        for (i, (pos, status)) in self.flushes.iter_mut().enumerate() {
+            if *pos == flush_pos {
+                *status = FlushStatus::Complete;
+                found_index = Some(i);
+            }
+            if !complete_prefix_finalized {
+                if let FlushStatus::Complete = status {
+                    complete_prefix_len += 1;
+                } else {
+                    complete_prefix_finalized = true;
+                }
             }
+            if found_index.is_some() && complete_prefix_finalized {
+                break;
+            }
+        }
+        found_index.expect("missing flush");
+        if complete_prefix_len > 0 {
+            let (pos, _) = self.flushes.drain(..complete_prefix_len).last().unwrap();
+            self.flushed_pos = pos;
         }
     }
 
     fn current_pending(&mut self) -> Vec<task::JoinHandle<()>> {
-        let mut handles = Vec::new();
-        for (_k, v) in self.pending.drain() {
-            handles.push(v);
+        let mut handles = Vec::with_capacity(self.pending_flush_count);
+        for (_, status) in &mut self.flushes {
+            if let FlushStatus::Pending(handle) = status {
+                if let Some(h) = handle.take() {
+                    handles.push(h);
+                }
+            }
         }
         handles
     }
 
     fn flush_one_buffer(&mut self, buffer: DmaBuffer, state: Rc<RefCell<Self>>, file: Rc<DmaFile>) {
         let aligned_pos = self.aligned_pos;
+        let flush_pos = self.current_pos();
         self.aligned_pos += self.buffer_size as u64;
         self.buffer_pos = 0;
-        let flush_id = self.flush_id;
-        self.flush_id += 1;
-        self.pending.insert(
-            flush_id,
-            Local::local(async move {
-                let res = file.write_at(buffer, aligned_pos).await;
-                let mut state = state.borrow_mut();
-                if !collect_error!(state, res) {
-                    state.adjust_flushed_pos(aligned_pos);
-                }
-                state.pending.remove(&flush_id); // can be None during close
-                if let Some(waker) = state.waker.take() {
-                    drop(state);
-                    waker.wake();
-                }
-            })
-            .detach(),
-        );
+        self.pending_flush_count += 1;
+        let handle = Local::local(async move {
+            let res = file.write_at(buffer, aligned_pos).await;
+            let mut state = state.borrow_mut();
+            state.pending_flush_count -= 1;
+            if !collect_error!(state, res) {
+                state.on_flush_success(flush_pos);
```

Comment:
This gives me pause. What should be the behavior on error? With the current state of this PR, we would end up with that failed `flush_pos` stuck as `FlushStatus::Pending`.